### PR TITLE
Implement deepcopy for BigInt and BigFloat

### PR DESCRIPTION
**Pull request by yuyichao**
Workaround #16667 for `BigFloat` and `BigInt`

I was hoping that someone that have actually used `gmp`/`mpfr` can implement this since it took me a while to figure out the right function that `copy`s the value (it was called `assignment` in the doc)

Replaces https://github.com/JuliaLang/julia/pull/16826

### CODE REVIEW

File: test/copy.jl [hunk 1]
```diff
@@ -82,3 +82,29 @@ let a1 = Base.svec(1, 2, 3, []), a2 = Base.svec(1, 2, 3)
...
+    # independent
+    for v in x
+        finalize(v)
+    end
+    # Allocate some memory to make it more likely to trigger an error
+    # if `deepcopy` went wrong
+    x = BigInt[1:1000;]
+    @test y == x
+end
+let x = BigFloat[1:1000;], y
```
Comment (rfourquet) [inline test/copy.jl]: the trailing `y` is intended?

File: base/mpfr.jl [hunk 2]
```diff
@@ -880,4 +880,18 @@ get_emin_max() = ccall((:mpfr_get_emin_max, :libmpfr), Clong, ())
...
 set_emin!(x) = ccall((:mpfr_set_emin, :libmpfr), Void, (Clong,), x)
 
+function Base.deepcopy_internal(x::BigFloat, stackdict::ObjectIdDict)
+    if haskey(stackdict, x)
+        return stackdict[x]
+    end
+    N = precision(x)
+    y = BigFloat(zero(Clong), zero(Cint), zero(Clong), C_NULL)
+    ccall((:mpfr_init2,:libmpfr), Void, (Ptr{BigFloat}, Clong), &y, N)
+    finalizer(y, Base.GMP._mpfr_clear_func)
```
Comment (rfourquet) [inline base/mpfr.jl]: I'm not familiar with mpfr, but any reason not to replace those 4 lines by `y = setprecision(precision(x)) do; BigFloat() end` ?

**Comment by rfourquet** [comment]
LGTM and +1, as I think the other fix by SPJ was a bit dangerous as it is unavoidable that some people will break over the officially asserted immutability of `BigInt/BigFloat`, for efficienty reasons (i.e. need a `deepcopy` of `x` and then mutate the copy using gmp/mpfr library functions).

File: test/copy.jl (hunk 1 above)
Comment (yuyichao) [inline test/copy.jl]: Yes, it makes sure y is also local, similar to the `y = ...` above.

File: base/mpfr.jl (hunk 2 above)
Comment (yuyichao) [inline base/mpfr.jl]: They are equivalent at MPFR level (i.e. we'll make identical calls) but this version has slightly less code at julia level (the `setprecision` version sets the default precision in julia)

**Comment by yuyichao** [comment]
> and then mutate the copy using gmp/mpfr library functions

That is still as undefined as anything else that relies on the internal of the type although I'm not sure what's the convention for big numbers (especially since we are wrapping a library with it's own convention). The case I want to fix is mainly the one in the test.

File: test/copy.jl (hunk 1 above)
Comment (rfourquet) [inline test/copy.jl]: Sorry for a possibly naive question, but then why not making sure `z` below is local too?

Comment (yuyichao) [inline test/copy.jl]: (Because it was added later ;-p) Will do.

File: base/mpfr.jl (hunk 2 above)
Comment (rfourquet) [inline base/mpfr.jl]: Oh I see, for maximum efficiency...!

File: test/copy.jl (hunk 1 above)
Comment (rfourquet) [inline test/copy.jl]: Impressive rigor, many tests elsewhere don't care at all about global namespace pollution!

**Comment by rfourquet** [comment]
Scott on julia-users:

> I'm concerned about the performance implications of PR #16999, which makes a copy any BigFloats or BigInts when doing deepcopy.
The rationale for doing so is that somebody might bypass the BigInt and BigFloat types, and directly operate on their GMP/MPFR internals.

> A few comments here (since I can't comment in-line at #16999).

> @rfourquet: LGTM and +1, as I think the other fix by SPJ was a bit dangerous as it is unavoidable that some people will break over the officially asserted immutability of BigInt/BigFloat, for efficienty reasons (i.e. need adeepcopy of x and then mutate the copy using gmp/mpfr library functions).

> If they are doing things for efficiency reasons, then they will not want to be unnecessarily doing copies of BigInt and BigFloat in this fashion, because this means that any time you are doing a deepcopy of an array (for example, so that you can sort it),
you end up 1) allocating a new Julia object 2) allocating memory managed by GMP/MPFR 3) adding a finalizer for every number copied.  I can do a benchmark later to show just how much that would cost (and can be prohibitive if you are dealing with millions of big numbers).

> Are there any examples in any of the packages of people directly operating on the internals of `BigInt`s or `BigFloat`s where this would be an issue?
There definitely are people who were using `deepcopy` because they needed the arrays copied, before the arrays were modified.
Now they would need to make their own "deepcopy", that only copied the containers, and not the elements which are supposed to be scalar numbers.

> As far as the dangerousness, there is the same issue with strings, which by convention are immutable, but by implementation are not.   If you started directly modifying the internal .data field, you would definitely break a lot of code.

> Unfortunately, there is no way currently of enforcing this type of invariant in Julia.

**Comment by rfourquet** [comment]
I don't have any example where this would be an issue, and honestly I didn't think to much about this problem, I just thought it looked safer (so I step back!). I myself have occasionally used `x+0` to get a "deep copy" of a BigInt x. But @yuyichao said anyway it was not the use case he wanted to address, but rather the finalise problem...
ps: to sort an array, a shallow copy should suffice, no?

**Comment by yuyichao** [comment]
> I'm concerned about the performance implications of PR #16999, which makes a copy any BigFloats or BigInts when doing deepcopy.

So the suggestion is that if an operation following the defined schematics is too expensive, we should just ignore what the function should do and do nothing instead.

> The rationale for doing so is that somebody might bypass the BigInt and BigFloat types, and directly operate on their GMP/MPFR internals.

No it's not.

**Comment by tkelman** [comment]
travis osx failure was missing llvm binaries, log backed up to https://gist.github.com/3582f97886aeb5c6e1471267db570ed8 and restarted

**Comment by rfourquet** [comment]
@ScottPJones  and @JeffreySarnoff replied there: https://groups.google.com/forum/#!topic/julia-users/EzMFkgKb4oY (not pasting here as there are nested comments).

**Comment by rfourquet** [comment]
I think Scott has a point, in that it's not necessarily clear why the gmp/mpfr have to be copied. I will speak as someone who may implement similar types and who needs to understand the precise rules for deepcopy.

 In the doc, it's said "deepcopy on an object should generally have the same effect as serializing and deserializing" (Stefan [adds](https://github.com/JuliaLang/julia/issues/16667#issuecomment-223047312) "except it's much more efficient"). I don't know much about serialization in Julia, but I can read there that "The read-back value will be as identical as possible to the original". In this respect, Scott's fix was the winner, as what can be more identical to the original than the original itself? 

And there is no mention of the "finalizer problem" in those docs (furthermore, "should generally" in the doc of deepcopy is too vague to convey the notion that it's necessary to address this finalizer problem). So it's not documented AFAIK that after calling `finalize` on a `deepcopy` of an object, it's safe to use the original object. And I don't understand why it should be safe, so I'm missing one piece of the puzzle, please someone document that!

**Comment by yuyichao** [comment]
> I think Scott has a point

No he doesn't and never did.

> who needs to understand the precise rules for deepcopy.

I'll never make the mistake of wasting my time arguing with SGJ again but I certainly agree that the doc should be improved and I'm happy to explain the doc to you here.

> generally have the same effect as serializing and deserializing
> I can read there that "The read-back value will be as identical as possible to the original".
> In this respect, Scott's fix was the winner, as what can be more identical to the original than the original itself?

Multiple logical fallacy here,

1. This (`what can be more identical to the original than the original itself`) is basically claiming that we should define `deepcopy(x) = x`, which is clearly wrong.
2. `as identical as possible` is the basic requirement for any kind of `copy`, and has nothing to do with the schematics of `deepcopy` or `copy` (in that it is not at all the factor to decide how each of them should be implemented). This should be clear since we don't implement either of them as identity for all objects.
3. The important part of `same effect as serializing and deserializing` is that they should be totally independent. "Serializing and deserializing" is roughly equal to save an object to disc, restart the julia session, and load it from disk again. The new object shouldn't share any information with the old one that depends on the environment the old object is defined that is not present in the new session. "Same effect" should at least mean that you shouldn't be able to tell the difference by doing any well defined operation on the two objects to notice that it is not actually serialized and deserialized. Both `is` and `finalize` are documented and well defined operations on these objects so you shouldn't be able to use them to observe the difference between the two approaches (serialize + deserialize v.s. deepcopy)

> And there is no mention of the "finalizer problem" in those docs

The fact that finalizer is a problem was clearly not carefully thought of before, which is why they are not mentioned in the doc. I'm not sure what should we mention before we fix #16667 though.

> So it's not documented AFAIK that after calling finalize on a deepcopy of an object, it's safe to use the original object. And I don't understand why it should be safe, so I'm missing one piece of the puzzle, please someone document that!

As mentioned above, `finalize` is a well defined function to call and it shouldn't be used to observe the difference from serialize+deserialize. It is certainly fine to mention this in the doc, I'm not sure if it is special enough though (maybe it is).

**Comment by JeffreySarnoff** [comment]
_replacing my comment in julia-users with this ... corrected and clarified:_

There is risk in decoupling from the use of computationally attributed terms their semantic intent.  That one can, by art or artifice, flip a bit within the state of an immutably typed variable does not change the sensible meaning of "immutable" to "episodic unchangingness" or "unalterable unless altered." 

> immutable NotMutable{T} val::T end;  nonmutagenic = NotMutable(BigFloat(pi)); 
> swizzleable  = sizeof(nonmutagenic) < sizeof(nonmutagenic.val) # true, 8 < 24
> 

The declaration of intentional design that is evident with use of _immutable_ must be stronger and more compelling than the fact that it is swizzleable.  Occurrences of an immutable type are `not` quantum entangled with barbed wire.  I suggest the considerations of  practice, performance and accommodation are less salient than the risk that allowing the immutable to mutate is decohering of Julia.

Julia's metaconfigurational strength and unfurling architectural flexibility are real and  substantive skills with which she now moves.   That there is collaborative adherence to collectively adopted ways to software wellness and that there is an ongoing pleasantness she brings into person-to-person communicative development of technological applications are some keys Julia always carries.

**Comment by rfourquet** [comment]
OK thanks for the extensive answer. I guess `BigInt` is an edge case as the desired semantics is to be considered an immutable (even if it's not technically), so obtaining "a fully independant object" can be interpreted at both levels. I just didn't take the "should generally have the same effect as serializing and deserializing" as an absolute strict requirement (and genuinely don't understand the use case of copying the data of `BigInt` when deep-copying).

**Comment by yuyichao** [comment]
> I guess BigInt is an edge case

If anything, it is not as much an edge case than `String`.

**Comment by rfourquet** [comment]
By saying `BigInt`, it was short for `BigInt`, `BigFloat` and `String` (and "the like"). I'm curious to know why `String` is more an edge case.

**Comment by yuyichao** [comment]
Because it doesn't have finalizer.

**Comment by rfourquet** [comment]
Oh indeed.

**Comment by StefanKarpinski** [comment]
@yuyichao: could you make the case more clearly for why you think that BigInt and BigFloat need to be copied by deepcopy? I can't find your argument made clearly anywhere.

**Comment by yuyichao** [comment]
Here? https://github.com/JuliaLang/julia/pull/16999#issuecomment-226980465

Short summary.

1. We want to define `deepcopy` to have the same effect as `serialize` and `deserialize`.
2. I interpret "same effect" as it is not possible to observe the difference using any well defined operations.
3. `is` and `finalize` are documented and well defined operations.
4. Therefore you cannot use `is` and `finalize` to observe that the result of `deepcopy`ing an object is not implemented as `serialize` and `deserialize`.
5. If `deepcopy` on a mutable type with finalizer is implemented as identity, you can use `is` to see that the new object is identical with the old one (which is never the case for `serialize` and `deserialize`) as well as using `finalize` to observe that the other object is corrupted when one of them is `finalize`d (also never the case for `serialize` and `deserialize`).
6. Therefore identity is not the right implementation.
7. This PR is just the most straightforward implementation I can come up with that satisfies the requirements. (Ref counting the GMP object can work too, for example.)

**Comment by yuyichao** [comment]
Also [here](https://github.com/JuliaLang/julia/pull/16826#issuecomment-224597982)

**Comment by mauro3** [comment]
Sorry, I'm out of my depth here, but my question: why can't BigFloat just be immutable?  If immutable, this would solve some of the issues?

**Comment by yuyichao** [comment]
immutable can't have finalizers

**Comment by yuyichao** [comment]
And in general, whether it's immutable or not is not the issue at all, the issue is that there's finalizer, which has to be independent for reasons shown above.

**Comment by rfourquet** [comment]
[disclaimer] I (a modest user who don't know much of Julia internals) don't pretend to know how things should be, but as there is some frustration for more that 1 person due to the inefficiency of this patch for certain (real) use cases, more clarifying is needed (in particular to help us understand the rules when we override `deepcopy` for user-defined types).

@yuyichao, I had more or less already understood your reasoning that you have summarized above. But the big elephant in the room is point 1: why should "we" want `deepcopy` to have the same effect as `serialize/deserialize`, to the point that no difference other than speed of operation should be observable (your point 2)? This is not stated as a guaranty in the doc of `deepcopy` (and there seems to be an exception for functions?). To the casual reader, this property (which "should generally" hold) appears merely as a guideline to convey roughly what should be the effect of `deepcopy`. Until a real use case shows otherwise, it cannot make sense to aim to be undistinguishable from `serialize/deserialize` for its own sake. My understanding is that `deepcopy` is meant to be a recursive copy (i.e. for collections, or types which have mutable fields). The official API of `BigInt` (or `String`, etc.) is to be immutable, and `finalize` is not part of it: if a user finalizes directly a `BigInt`, or mutates it's internal structure, the API boundary has been crossed, and it's fair to not guaranty anything in this case. On the other hand, a "normal" user is not supposed to depend on the implementation details of `BigInt`, and so never needs to have `deepcopy(x::BigInt)` return something else than `identity(x)` i.e. `x`: there are no officialy supported tools to take advantage of having a copy (at the memory level). It is not obvious why this user has to pay the price of copying the data only to support an extremely obscure use case where someone would finalize directly a `BigInt`. This position came to me after you rightly corrected me (above) on the mutating aspect: "That is still as undefined as anything else that relies on the internal of the type". I just don't see how relying on being allowed to finalize a `deepcopy` of a a `BigInt` and still using the original is not "relying on the internals".

In conclusion, until I understand the real problem solved by this patch, I consider that a better solution is to return the object itself, and possibly to document this exception in the `deepcopy` docstring.

I apologize in advance for any logical fallacy in my reasoning, be it real or perceived due to the inadequacy and imprecision of my words, it would be genuine.

PS: I don't perceive clearly what you think should do `deepcopy` on `String` (which currently copies the buffer, but no finalizer is involved).

**Comment by yuyichao** [comment]
`deepcopy` is a way to get `a fully independent object`, which is why it is defined as `have the same effect as serializing and then deserializing it`. Changing the definition is not part of this PR.

`deepcopy` is **NOT** defined as a way to unshare all the containers and not anything the caller happens to be not interested in. Satisfying all those requirement would be impossible with a single function. And for those cases defining a custom copy (when necessary, note that ODE.jl should just use `copy` instead) is definitely faster in all cases.

**Comment by mauro3** [comment]
Concerning ODE.jl, say if a user uses a state vector of type `typeof(y0)==Vector{Matrix{Float64}}` and I would like to store a complete copy of this (for the output), then a deep-copy is needed. So, to keep the code generic, using `deepcopy` seems correct, even though this is not needed for all types.  Thus, ideally for ODE.jl, for the types which don't need it, `deepcopy` would be as efficient as `copy`.  Anyway, I should probably just benchmark it and see whether special-casing BigFloat (and other types) would make sense.  Thanks!

**Comment by rfourquet** [comment]
I believe for `BigInt` etc. there is no need of special case, but rather of a relaxed definition and possibly a clearer guideline. There is no need to remove the mention of serialize and deserialize, as it is not stated as a strict requirement.

**Comment by yuyichao** [comment]
The serialization and deserializeis how independent is defined for deepcopy. So you are proposing a special case unless the definition is changed

**Comment by tknopp** [comment]
@rfourquet: Could you elaborate what the use case is that frustrates more than one person so much? Where is a `deepcopy` of a `BigInt` array needed?

**Comment by rfourquet** [comment]
Yes I was precisely proposing a "relaxed definition", which currently have anyhow some amount of ambiguity (if only because we both don't interpret the docstring in the same way). 
@tknopp a user mentioned its use case on gitter; what was needed was not a `deepcopy` of a `BigInt`, but a deepcopy of a structure which contains `BigInt` at more that one level of depth (e.g. arrays of arrays of `BigInt`).

**Comment by tknopp** [comment]
@rfourquet: This does not answer my question. It would be good if "that user" would open a dedicated issue that reveals the performance issue and has a short use case scenario. Then one can long term track the feature request of a fast `deepcopy`. There are various performance issues in the bug tracker of julia and since @yuyichao's patch here fixed an actual failure it would would be good to acknowledge this and leave performance optimizations to the other issue.

**Comment by rfourquet** [comment]
@tknopp yes sorry, I didn't expand more here on that user's performance problem, in part because as you said this is not the right place to continue the discussion. My intention is to think more about this when I have the bandwidth and possibly open a new PR. But this was still IMHO acceptable to discuss the performance implications of this PR, as there was another PR solving the same issue; I needed to understand why this PR would be chosen over the other one. I have no problem aknowledging the useful contribution of this patch which is on the safest side.

**Comment by tkelman** [comment]
I think this is giving a double-free when I try to backport it to release-0.4
```
julia> using Base.Test

julia> let x = BigInt[1:1000;], y = deepcopy(x), v
           # Finalize the original values to make sure the deep copy is indeed
           # independent
           for v in x
               finalize(v)
           end
           # Allocate some memory to make it more likely to trigger an error
           # if `deepcopy` went wrong
           x = BigInt[1:1000;]
           @test y == x
       end
warning: Critical error detected c0000374

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffe0729a6f3 in ntdll!RtlpNtMakeTemporaryKey ()
   from /cygdrive/c/WINDOWS/SYSTEM32/ntdll.dll
(gdb) bt
#0  0x00007ffe0729a6f3 in ntdll!RtlpNtMakeTemporaryKey ()
   from /cygdrive/c/WINDOWS/SYSTEM32/ntdll.dll
#1  0x00007ffe0729859a in ntdll!RtlpNtMakeTemporaryKey ()
   from /cygdrive/c/WINDOWS/SYSTEM32/ntdll.dll
#2  0x00007ffe07245d86 in ntdll!RtlRaiseStatus ()
   from /cygdrive/c/WINDOWS/SYSTEM32/ntdll.dll
#3  0x00007ffe071dcb19 in ntdll!RtlFreeHeap ()
   from /cygdrive/c/WINDOWS/SYSTEM32/ntdll.dll
#4  0x00007ffe047d9dcc in msvcrt!free () from /cygdrive/c/WINDOWS/System32/msvcrt.dll
#5  0x000000006d10f386 in jl_gc_counted_free (p=<optimized out>, sz=631064704)
    at gc.c:2555
#6  0x000000006d1048d4 in jl_apply (nargs=1, args=0xd6f4f0, f=0x6e0b1d18 <jl_pgcstack>)
    at julia.h:1331
#7  run_finalizer (o=0x224ca100, o@entry=0x83bd6650, ff=0x6e0b1d18 <jl_pgcstack>)
    at gc.c:324
#8  0x000000006d10abab in finalize_object (list=0x6e0b4240 <finalizer_list>,
    list=0x6e0b4240 <finalizer_list>, o=0x83bd6650) at gc.c:347
#9  jl_finalize (o=0x83bd6650) at gc.c:406
#10 0x00000000054c223a in ?? ()
```

**Comment by yuyichao** [comment]
The fix isn't the issue. The test happens to trigger another bug on 0.4 (`finalize` doesn't handle C function finalizer), which is arguably more urgent than this one, I'll try to find the commit/pr that fixes it later.

**Comment by yuyichao** [comment]
@tkelman The bug is https://github.com/JuliaLang/julia/issues/15408, see the commits referenced in it for the backportable fix and related tests.

**Comment by tkelman** [comment]
Right, I was looking at the test for that in an end-of-file conflict on some other backport. How badly do we need both on release-0.4?

**Comment by yuyichao** [comment]
Both are fixing segfault for dealing with unsafe (C) code so they are important fixes. OTOH, no one complained about #15408 on 0.4 so I'm not sure how important that is in practice.

**Comment by JeffreySarnoff** [comment]
yes -- the segfault thing happens sometimes
